File: ios/loveaceios/BLE/BLECommandHelper.swift
import Foundation

enum BLECommandHelper {
    static let serviceUUID = "0000FFE0-0000-1000-8000-00805F9B34FB"
    static let characteristicUUID = "0000FFE1-0000-1000-8000-00805F9B34FB"

    static func bcc(_ hex: String) -> String {
        guard !hex.isEmpty else { return "00" }
        let len = hex.count / 2
        var result = 0
        for i in 0..<len {
            let start = hex.index(hex.startIndex, offsetBy: i * 2)
            let end = hex.index(start, offsetBy: 2)
            let byteStr = String(hex[start..<end])
            result ^= Int(byteStr, radix: 16) ?? 0
        }
        return String(format: "%02X", result)
    }

    static func buildOpenDoorCommand(sKey: String, schoolId: String) -> Data {
        let opCode = schoolId == "340301" ? "05FF" : "0503"
        let cmd = opCode + sKey
        let full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    static func buildAddCardCommand(cardId: String, sn: Int, sKey: String, endDateTime: String) -> Data {
        let snHex = String(format: "%04d", sn)
        let cmd = "1007\(cardId)\(snHex)\(sKey)\(endDateTime)"
        let full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    static func buildFreezeCardCommand(sn: Int, sKey: String, endDateTime: String) -> Data {
        let snHex = String(format: "%04d", sn)
        let cmd = "100700000000\(snHex)\(sKey)\(endDateTime)"
        let full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    static func buildCheckTimeCommand() -> Data {
        let bcdTime = bleTimeBCD()
        let cmd = "0810\(bcdTime)"
        let full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    static func buildAlwaysOpenCommand(sKey: String, schoolId: String) -> Data {
        let opCode = schoolId == "340301" ? "05FF" : "0511"
        let cmd = opCode + sKey
        let full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    static func buildAlwaysOffCommand(sKey: String, schoolId: String) -> Data {
        let opCode = schoolId == "340301" ? "05FF" : "0512"
        let cmd = opCode + sKey
        let full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    private static func bleTimeBCD() -> String {
        let cal = Calendar.current
        let now = Date()
        let y = max(cal.component(.year, from: now) - 2000, 0)
        let m = cal.component(.month, from: now)
        let d = cal.component(.day, from: now)
        let h = cal.component(.hour, from: now)
        let min = cal.component(.minute, from: now)
        let s = cal.component(.second, from: now)
        return String(format: "%02d%02d%02d%02d%02d%02d", y, m, d, h, min, s)
    }

    static func hexToBytes(_ hex: String) -> Data {
        var data = Data()
        let len = hex.count / 2
        for i in 0..<len {
            let start = hex.index(hex.startIndex, offsetBy: i * 2)
            let end = hex.index(start, offsetBy: 2)
            if let byte = UInt8(hex[start..<end], radix: 16) { data.append(byte) }
        }
        return data
    }

    static func bytesToHex(_ data: Data) -> String {
        data.map { String(format: "%02x", $0) }.joined()
    }

    static func parseOpenDoorResponse(_ hex: String) -> (recordNum: Int, power: Int)? {
        let lower = hex.lowercased()
        guard let idx = lower.range(of: "0703")?.lowerBound else { return nil }
        let startOffset = lower.distance(from: lower.startIndex, to: idx) + 4
        guard startOffset + 8 <= hex.count else { return nil }
        let numStart = hex.index(hex.startIndex, offsetBy: startOffset)
        let numEnd = hex.index(numStart, offsetBy: 4)
        let powerEnd = hex.index(numEnd, offsetBy: 4)
        guard let num = Int(hex[numStart..<numEnd], radix: 16),
              let power = Int(hex[numEnd..<powerEnd], radix: 16) else { return nil }
        return (num, power)
    }
}

File: android/app/src/main/java/tech/loveace/appv3/data/ble/BleCommandHelper.kt
package tech.loveace.appv3.data.ble

import java.util.Calendar

/**
 * BLE 门锁指令构建工具 — 移植自小程序 utilsStr.js / ble.js
 */
object BleCommandHelper {

    // BLE 服务和特征 UUID
    const val SERVICE_UUID = "0000FFE0-0000-1000-8000-00805F9B34FB"
    const val CHARACTERISTIC_UUID = "0000FFE1-0000-1000-8000-00805F9B34FB"

    /** BCC 校验 */
    fun bcc(hex: String): String {
        if (hex.isEmpty()) return "00"
        val len = hex.length / 2
        var result = 0
        for (i in 0 until len) {
            result = result xor hex.substring(i * 2, i * 2 + 2).toInt(16)
        }
        return result.toString(16).uppercase().padStart(2, '0')
    }

    /** 开门指令 */
    fun buildOpenDoorCommand(sKey: String, schoolId: String): ByteArray {
        val opCode = if (schoolId == "340301") "05FF" else "0503"
        val cmd = opCode + sKey
        val full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    /** 发卡指令 */
    fun buildAddCardCommand(cardId: String, sn: Int, sKey: String, endDateTime: String): ByteArray {
        val snHex = sn.toString().padStart(4, '0')
        val cmd = "1007$cardId$snHex$sKey$endDateTime"
        val full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    /** 冻结卡指令 */
    fun buildFreezeCardCommand(sn: Int, sKey: String, endDateTime: String): ByteArray {
        val snHex = sn.toString().padStart(4, '0')
        val cmd = "100700000000${snHex}${sKey}$endDateTime"
        val full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    /** 校时指令 */
    fun buildCheckTimeCommand(): ByteArray {
        val bcdTime = bleTimeBCD()
        val cmd = "0810$bcdTime"
        val full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    /** 常开指令 */
    fun buildAlwaysOpenCommand(sKey: String, schoolId: String): ByteArray {
        val opCode = if (schoolId == "340301") "05FF" else "0511"
        val cmd = opCode + sKey
        val full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    /** 常闭指令 */
    fun buildAlwaysOffCommand(sKey: String, schoolId: String): ByteArray {
        val opCode = if (schoolId == "340301") "05FF" else "0512"
        val cmd = opCode + sKey
        val full = cmd + bcc(cmd)
        return hexToBytes(full)
    }

    /** BCD 时间编码 */
    private fun bleTimeBCD(): String {
        val cal = Calendar.getInstance()
        val y = (cal.get(Calendar.YEAR) - 2000).coerceAtLeast(0)
        val m = cal.get(Calendar.MONTH) + 1
        val d = cal.get(Calendar.DAY_OF_MONTH)
        val h = cal.get(Calendar.HOUR_OF_DAY)
        val min = cal.get(Calendar.MINUTE)
        val s = cal.get(Calendar.SECOND)
        return "%02d%02d%02d%02d%02d%02d".format(y, m, d, h, min, s)
    }

    /** Hex 字符串转 ByteArray */
    fun hexToBytes(hex: String): ByteArray {
        val len = hex.length / 2
        val bytes = ByteArray(len)
        for (i in 0 until len) {
            bytes[i] = hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
        return bytes
    }

    /** ByteArray 转 Hex 字符串 */
    fun bytesToHex(bytes: ByteArray): String {
        return bytes.joinToString("") { "%02x".format(it) }
    }

    /** 解析开门响应：返回 (recordNum, power) 或 null */
    fun parseOpenDoorResponse(hex: String): Pair<Int, Int>? {
        if (hex.contains("0703", ignoreCase = true)) {
            val idx = hex.lowercase().indexOf("0703")
            if (idx + 14 <= hex.length) {
                val numHex = hex.substring(idx + 4, idx + 8)
                val powerHex = hex.substring(idx + 8, idx + 12)
                val num = numHex.toIntOrNull(16) ?: return null
                val power = powerHex.toIntOrNull(16) ?: return null
                return Pair(num, power)
            }
        }
        return null
    }
}
